internal/controller/sync.go

```diff
@@ -207,159 +205,88 @@ func (r *Route) hasValidCertificate(route *routev1.Route) bool {
-		_, err := utilpki.DecodePrivateKeyBytes([]byte(route.Annotations[cmapi.IsNextPrivateKeySecretLabelKey]))
-		if err != nil {
-			r.eventRecorder.Event(route, corev1.EventTypeWarning, ReasonInvalidKey, "Regenerating Next Private Key as the existing key is invalid: "+err.Error())
-			return false
-		}
-		return true
-	}
-	return false
-}
-
-func (r *Route) generateNextPrivateKey(ctx context.Context, route *routev1.Route) error {
-	privateKeyAlgorithm, found := route.Annotations[cmapi.PrivateKeyAlgorithmAnnotationKey]
-	if !found {
-		privateKeyAlgorithm = string(cmapi.RSAKeyAlgorithm)
-	}
-
-	var privateKeySize int
-	privateKeySizeStr, found := route.Annotations[cmapi.PrivateKeySizeAnnotationKey]
-	if found {
-		var err error
-		privateKeySize, err = strconv.Atoi(privateKeySizeStr)
-		if err != nil {
-			r.eventRecorder.Event(route, corev1.EventTypeWarning, ReasonInvalidPrivateKeySize, "invalid private key size:"+privateKeySizeStr)
-			return fmt.Errorf("invalid private key size, %s: %v", privateKeySizeStr, err)
-		}
-	} else {
-		switch privateKeyAlgorithm {
-		case string(cmapi.ECDSAKeyAlgorithm):
-			privateKeySize = utilpki.ECCurve256
-		case string(cmapi.RSAKeyAlgorithm):
-			privateKeySize = utilpki.MinRSAKeySize
-		}
-	}
-
-	var privateKey crypto.PrivateKey
-	var err error
-	switch privateKeyAlgorithm {
-	case string(cmapi.ECDSAKeyAlgorithm):
-		privateKey, err = utilpki.GenerateECPrivateKey(privateKeySize)
-		if err != nil {
-			return fmt.Errorf("could not generate ECDSA key: %w", err)
-		}
-	case string(cmapi.RSAKeyAlgorithm):
-		privateKey, err = utilpki.GenerateRSAPrivateKey(privateKeySize)
-		if err != nil {
-			return fmt.Errorf("could not generate RSA Key: %w", err)
-		}
-	default:
-		r.eventRecorder.Event(route, corev1.EventTypeWarning, ReasonInvalidPrivateKeyAlgorithm, "invalid private key algorithm: "+privateKeyAlgorithm)
-		return fmt.Errorf("invalid private key algorithm: %s", privateKeyAlgorithm)
-	}
-	encodedKey, err := utilpki.EncodePrivateKey(privateKey, cmapi.PrivateKeyEncoding(cmapi.PKCS1))
-	if err != nil {
-		return fmt.Errorf("could not encode %s key: %w", privateKeyAlgorithm, err)
-	}
-	route.Annotations[cmapi.IsNextPrivateKeySecretLabelKey] = string(encodedKey)
-	_, err = r.routeClient.RouteV1().Routes(route.Namespace).Update(ctx, route, metav1.UpdateOptions{})
-	if err != nil {
-		return err
-	}
-	r.eventRecorder.Event(route, corev1.EventTypeNormal, ReasonIssuing, "Generated Private Key for route")
-	return nil
-}
-
-func getCurrentRevision(route *routev1.Route) (int, error) {
-	revision, found := route.Annotations[cmapi.CertificateRequestRevisionAnnotationKey]
-	if !found {
-		return 0, fmt.Errorf("no revision found")
-	}
-	return strconv.Atoi(revision)
-}
+func (r *RouteController) getCertificateForRoute(ctx context.Context, route *routev1.Route) (*cmapi.Certificate, error) {
+	// Note: this could also implement logic to re-use an existing certificate: route.Annotations[cmapi.CertificateNameKey]
```

Comment:
Does this mean that it isn't possible (yet) to pre-create a Certificate in order to use a Certificate field that isn't supported using Route annotations?

(mentioned by @ctrought in https://github.com/cert-manager/openshift-routes/issues/49#issuecomment-1973483925)

Reply:
That's correct as far as I know yeah. That sounds like a nice-to-have rather than a need-to-have; do you agree?

I think a more pressing need is what was briefly discussed in standup yesterday, which is that this logic mirrors so much about ingress-shim that we could expose the ingress-shim logic from cert-manager and re-use it here.  But I didn't want to bite off a cert-manager change here.

Something like (pseudo function) `cmingress.AnnotationsToCertificate(...)`.

Reply:
I agree, let's postpone this to later (if folks still need it).

> this logic mirrors so much about ingress-shim that we could expose the ingress-shim

It would be great if we could do that, especially since we aim to support all the same annotations. 

Is it possible do this this now (with no work needed on the cert-manager project), or does it require changes in the cert-manager project in which case we would probably be better off doing that in the next release of openshift-routes?